### DISCUSSION TITLE
electron-chromedriver installation failure

### DISCUSSION THREAD

**Discussion by RagipSelcuk** [eclipse-theia/theia#12553] [Show and tell] [answered]
Hi folks,

 I have encountered below issue while "yarn" process is ongoing at Blueprint. As well as, I have successfully using the theia platform.

I did remove the node modules directory and running the "yarn" command, unfortunately result is the same.
I have tried on Windows PC and Mac Pc . Am I missing a big step as well as known by all developers? I'm not sure.

I'd be appreciated to any response for find a way.

```
ragipselcuk@Ragips-MacBook-Air theia-blueprint % yarn
yarn install v1.22.19
[1/5] 🔍  Validating package.json...

[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
[4/5] 🔗  Linking dependencies...
warning "@typescript-eslint/eslint-plugin > tsutils@3.21.0" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
warning " > @typescript-eslint/eslint-plugin-tslint@4.33.0" has unmet peer dependency "tslint@^5.0.0 || ^6.0.0".
warning " > @typescript-eslint/eslint-plugin-tslint@4.33.0" has unmet peer dependency "typescript@*".
warning " > eslint-plugin-deprecation@1.2.1" has unmet peer dependency "typescript@^3.7.5 || ^4.0.0".
warning " > ts-node@10.9.1" has unmet peer dependency "@types/node@*".
warning " > ts-node@10.9.1" has unmet peer dependency "typescript@>=2.7".
warning " > theia-blueprint-product@1.37.0" has unmet peer dependency "react@^16.8.0".
warning "workspace-aggregator-f4f8f1df-5923-4e63-9544-39f632f5a20e > theia-blueprint > wdio-chromedriver-service@6.0.4" has unmet peer dependency "chromedriver@*".
warning "workspace-aggregator-f4f8f1df-5923-4e63-9544-39f632f5a20e > theia-blueprint-browser > @theia/messages > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react@>=16.3.3".
warning "workspace-aggregator-f4f8f1df-5923-4e63-9544-39f632f5a20e > theia-blueprint-browser > @theia/messages > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react-dom@>=16.3.3".
warning "workspace-aggregator-f4f8f1df-5923-4e63-9544-39f632f5a20e > theia-blueprint-browser > @theia/process > node-pty@0.11.0-beta17" has unmet peer dependency "node-gyp@^8.3.0".
warning "workspace-aggregator-f4f8f1df-5923-4e63-9544-39f632f5a20e > theia-blueprint-browser > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react@^0.14.0 || ^15.0.0 || ^16.0.0".
warning "workspace-aggregator-f4f8f1df-5923-4e63-9544-39f632f5a20e > theia-blueprint-browser > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react-dom@^0.14.0 || ^15.0.0 || ^16.0.0".
[5/5] 🔨  Building fresh packages...
[-/14] ⠈ waiting...
[14/14] ⠈ electron-chromedriver
[-/14] ⠈ waiting...
[-/14] ⠈ waiting...
error /Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver: Command failed.
Exit code: 1
Command: node ./download-chromedriver.js
Arguments: 
Directory: /Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver
Output:
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
```

**Comment by vince-fugnitto** [comment]
@RagipSelcuk thank you for the discussion, I haven't noticed or encountered the issue before. 
Am I correct that you did the following:

-  fresh clone: `git clone git@github.com:eclipse-theia/theia-blueprint.git`
- move to the `theia-blueprint` directory
- perform `yarn`

The following works correctly for me, I tried both on Linux and MacOS.
I'm using `node` at version `v16.14.0`.

**Reply by RagipSelcuk** [reply]
#12329  @vince-fugnitto ,

 Many thanks for your response. You are right, I have followed as you figured.

Let me introduce configuration details.
<img width="853" alt="Screenshot 2023-05-24 at 00 23 15" src="https://github.com/eclipse-theia/theia/assets/6311780/fb599cde-22be-45bb-ae59-f75369866f0e">

I use MacOs Ventura Version 13.3.1 (a)

Please, you can find the command set which I used for installation.

#git clone https://github.com/eclipse-theia/theia-blueprint.git
#cd theia-blueprint

You can find the complete output of this steps, thanks in advance.

`ragipselcuk@Ragips-MacBook-Air blueprint % cd theia-blueprint 
ragipselcuk@Ragips-MacBook-Air theia-blueprint % yarn
yarn install v1.22.19
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
[4/5] 🔗  Linking dependencies...
warning "@typescript-eslint/eslint-plugin > tsutils@3.21.0" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
warning " > @typescript-eslint/eslint-plugin-tslint@4.33.0" has unmet peer dependency "tslint@^5.0.0 || ^6.0.0".
warning " > @typescript-eslint/eslint-plugin-tslint@4.33.0" has unmet peer dependency "typescript@*".
warning " > eslint-plugin-deprecation@1.2.1" has unmet peer dependency "typescript@^3.7.5 || ^4.0.0".
warning " > ts-node@10.9.1" has unmet peer dependency "@types/node@*".
warning " > ts-node@10.9.1" has unmet peer dependency "typescript@>=2.7".
warning " > theia-blueprint-product@1.37.0" has unmet peer dependency "react@^16.8.0".
warning "workspace-aggregator-a5533b3a-c261-4c1c-9a59-98d7662e3aa0 > theia-blueprint > wdio-chromedriver-service@6.0.4" has unmet peer dependency "chromedriver@*".
warning "workspace-aggregator-a5533b3a-c261-4c1c-9a59-98d7662e3aa0 > theia-blueprint-browser > @theia/messages > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react@>=16.3.3".
warning "workspace-aggregator-a5533b3a-c261-4c1c-9a59-98d7662e3aa0 > theia-blueprint-browser > @theia/messages > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react-dom@>=16.3.3".
warning "workspace-aggregator-a5533b3a-c261-4c1c-9a59-98d7662e3aa0 > theia-blueprint-browser > @theia/process > node-pty@0.11.0-beta17" has unmet peer dependency "node-gyp@^8.3.0".
warning "workspace-aggregator-a5533b3a-c261-4c1c-9a59-98d7662e3aa0 > theia-blueprint-browser > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react@^0.14.0 || ^15.0.0 || ^16.0.0".
warning "workspace-aggregator-a5533b3a-c261-4c1c-9a59-98d7662e3aa0 > theia-blueprint-browser > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react-dom@^0.14.0 || ^15.0.0 || ^16.0.0".
[5/5] 🔨  Building fresh packages...
[-/14] ⡀ waiting...
[-/14] ⡀ waiting...
[14/14] ⡀ electron-chromedriver
[-/14] ⡀ waiting...
error /Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver: Command failed.
Exit code: 1
Command: node ./download-chromedriver.js
Arguments: 
Directory: /Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver
Output:
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

GotError [HTTPError]: Response code 404 (Not Found) for https://github.com/electron/electron/releases/download/v9.0.0/chromedriver-v9.0.0-darwin-arm64.zip
    at EventEmitter.<anonymous> (/Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver/node_modules/got/source/as-stream.js:35:24)
    at EventEmitter.emit (node:events:513:28)
    at module.exports (/Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver/node_modules/got/source/get-response.js:22:10)
    at ClientRequest.handleResponse (/Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver/node_modules/got/source/request-as-event-emitter.js:155:5)
    at Object.onceWrapper (node:events:628:26)
    at ClientRequest.emit (node:events:525:35)
    at ClientRequest.origin.emit (/Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver/node_modules/@szmarczak/http-timer/source/index.js:37:11)
    at HTTPParser.parserOnIncomingClient [as onIncoming] (node:_http_client:693:27)
    at HTTPParser.parserOnHeadersComplete (node:_http_common:128:17)
    at TLSSocket.socketOnData (node:_http_client:534:22) {
  host: 'github.com',
  hostname: 'github.com',
  method: 'GET',
  path: '/electron/electron/releases/download/v9.0.0/chromedriver-v9.0.0-darwin-arm64.zip',
  socketPath: undefined,
  protocol: 'https:',
  url: 'https://github.com/electron/electron/releases/download/v9.0.0/chromedriver-v9.0.0-darwin-arm64.zip',
  gotOptions: {
    path: '/electron/electron/releases/download/v9.0.0/chromedriver-v9.0.0-darwin-arm64.zip',
    protocol: 'https:',
    slashes: true,
    auth: null,
    host: 'github.com',
    port: null,
    hostname: 'github.com',
    hash: null,
    search: null,
    query: null,
    pathname: '/electron/electron/releases/download/v9.0.0/chromedriver-v9.0.0-darwin-arm64.zip',
    href: 'https://github.com/electron/electron/releases/download/v9.0.0/chromedriver-v9.0.0-darwin-arm64.zip',
    retry: {
      retries: [Function (anonymous)],
      methods: Set(6) { 'GET', 'PUT', 'HEAD', 'DELETE', 'OPTIONS', 'TRACE' },
      statusCodes: Set(7) { 408, 413, 429, 500, 502, 503, 504 },
      errorCodes: Set(8) {
        'ETIMEDOUT',
        'ECONNRESET',
        'EADDRINUSE',
        'ECONNREFUSED',
        'EPIPE',
        'ENOTFOUND',
        'ENETUNREACH',
        'EAI_AGAIN'
      }
    },
    headers: {
      'user-agent': 'got/9.6.0 (https://github.com/sindresorhus/got)',
      'accept-encoding': 'gzip, deflate'
    },
    hooks: {
      beforeRequest: [],
      beforeRedirect: [],
      beforeRetry: [],
      afterResponse: [],
      beforeError: [],
      init: []
    },
    decompress: true,
    throwHttpErrors: true,
    followRedirect: true,
    stream: true,
    form: false,
    json: false,
    cache: false,
    useElectronNet: false,
    method: 'GET'
  },
  statusCode: 404,
  statusMessage: 'Not Found',
  headers: {
    server: 'GitHub.com',
    date: 'Tue, 23 May 2023 21:31:00 GMT',
    'content-type': 'text/plain; charset=utf-8',
    vary: 'X-PJAX, X-PJAX-Container, Turbo-Visit, Turbo-Frame, Accept-Encoding, Accept, X-Requested-With',
    'cache-control': 'no-cache',
    'strict-transport-security': 'max-age=31536000; includeSubdomains; preload',
    'x-frame-options': 'deny',
    'x-content-type-options': 'nosniff',
    'x-xss-protection': '0',
    'referrer-policy': 'no-referrer-when-downgrade',
    'content-security-policy': "default-src 'none'; base-uri 'self'; connect-src 'self'; form-action 'self'; img-src 'self' data:; script-src 'self'; style-src 'unsafe-inline'",
    'content-encoding': 'gzip',
    'transfer-encoding': 'chunked',
    'x-github-request-id': '5929:9A99:19755E33:19D8C29A:646D3094',
    connection: 'close'




ragipselcuk@Ragips-MacBook-Air theia-blueprint % `

**Reply by vince-fugnitto** [reply]
@RagipSelcuk are you perhaps behind a proxy? 

- https://github.com/electron/chromedriver/issues/42

**Reply by RagipSelcuk** [reply]
@vince-fugnitto I double checked the proxy setting but it is a little bit complicated.

When I check the output of the yarn, it doesn't get response from version of the electron driver.
"https://github.com/electron/electron/releases/download/v9.0.0/chromedriver-v9.0.0-darwin-arm64.zip"

I checked of the chrome driver V9.0.0 and I could not see the 'chromedriver-v9.0.0-darwin-arm64.zip' suffix.
V9.0.0 only includes below file.
[chromedriver-v9.0.0-darwin-x64.zip](https://github.com/electron/electron/releases/download/v9.0.0/chromedriver-v9.0.0-darwin-x64.zip)

is there any way to change of the yarn progress of url definition?

By the way, the latest version of the electron has the correct name convention of the "-darwin-arm64.zip" suffix. 
[chromedriver-v24.3.1-darwin-arm64.zip](https://github.com/electron/electron/releases/download/v24.3.1/chromedriver-v24.3.1-darwin-arm64.zip)


Regards.

**Reply by vince-fugnitto** [reply]
Do you mean what they express in the readme? https://github.com/electron/chromedriver#custom-mirror

**Reply by RagipSelcuk** [reply]
Hi again,

 Yes, I mean this approach. But unfortunately it is not solved . I have still same situation. 

I followed the below steps one by one.

1. I changed the "version:24.3.1" instead of "version:9.0.0" in package.json file at /theia-blueprint/node_modules/electron-chromedriver location.

```javascript
{
  "name": "electron-chromedriver",
  "version": "24.3.1",
  "description": "Electron ChromeDriver",
  "repository": "https://github.com/electron/chromedriver",
  "bin": {
    "chromedriver": "./chromedriver.js"
  },
  "scripts": {
    "install": "node ./download-chromedriver.js",
    "lint": "standard --fix",
    "test": "standard && mocha"
  },
  "engines": {
    "node": ">=10.12.0"
  },
  "author": "Kevin Sawicki",
  "license": "MIT",
  "dependencies": {
    "@electron/get": "^1.12.2",
    "extract-zip": "^2.0.0"
  },
  "devDependencies": {
    "mocha": "^7.1.2",
    "standard": "^13.1.0"
  }
}

``` 

2. I succesfully installed the "version:24.3.1" . As I said previous comments "version:9.0.0" does not include "-darwin-arm64.zip"

```javascript
ragipselcuk@Ragips-MacBook-Air electron-chromedriver % yarn install
yarn install v1.22.19
info No lockfile found.
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
warning mocha > debug@3.2.6: Debug versions >=3.2.0 <3.2.7 || >=4 <4.3.1 have a low-severity ReDos regression when used in a Node.js environment. It is recommended you upgrade to 3.2.7 or 4.3.1. (https://github.com/visionmedia/debug/issues/797)
warning mocha > chokidar > fsevents@2.1.3: "Please update to latest v2.3 or v2.2"
[3/5] 🚚  Fetching packages...
[4/5] 🔗  Linking dependencies...
[5/5] 🔨  Building fresh packages...
success Saved lockfile.
$ node ./download-chromedriver.js
✨  Done in 21.86s.
ragipselcuk@Ragips-MacBook-Air electron-chromedriver % 
``` 

3. After installation is completed. I turn back to top directory and typed the "yarn" but yarn working with fresh packages and rollback again to the "version:9.0.0" 

```javascript
ragipselcuk@Ragips-MacBook-Air theia-blueprint % yarn 
yarn install v1.22.19
[1/5] 🔍  Validating package.json...
[2/5] 🔍  Resolving packages...
[3/5] 🚚  Fetching packages...
[4/5] 🔗  Linking dependencies...
warning "@typescript-eslint/eslint-plugin > tsutils@3.21.0" has unmet peer dependency "typescript@>=2.8.0 || >= 3.2.0-dev || >= 3.3.0-dev || >= 3.4.0-dev || >= 3.5.0-dev || >= 3.6.0-dev || >= 3.6.0-beta || >= 3.7.0-dev || >= 3.7.0-beta".
warning " > @typescript-eslint/eslint-plugin-tslint@4.33.0" has unmet peer dependency "tslint@^5.0.0 || ^6.0.0".
warning " > @typescript-eslint/eslint-plugin-tslint@4.33.0" has unmet peer dependency "typescript@*".
warning " > eslint-plugin-deprecation@1.2.1" has unmet peer dependency "typescript@^3.7.5 || ^4.0.0".
warning " > ts-node@10.9.1" has unmet peer dependency "@types/node@*".
warning " > ts-node@10.9.1" has unmet peer dependency "typescript@>=2.7".
warning " > theia-blueprint-product@1.37.0" has unmet peer dependency "react@^16.8.0".
warning "workspace-aggregator-b3340927-bf1a-491e-bf4a-a3ffec7e9fd4 > theia-blueprint > wdio-chromedriver-service@6.0.4" has unmet peer dependency "chromedriver@*".
warning "workspace-aggregator-b3340927-bf1a-491e-bf4a-a3ffec7e9fd4 > theia-blueprint-browser > @theia/messages > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react@>=16.3.3".
warning "workspace-aggregator-b3340927-bf1a-491e-bf4a-a3ffec7e9fd4 > theia-blueprint-browser > @theia/messages > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react-dom@>=16.3.3".
warning "workspace-aggregator-b3340927-bf1a-491e-bf4a-a3ffec7e9fd4 > theia-blueprint-browser > @theia/process > node-pty@0.11.0-beta17" has unmet peer dependency "node-gyp@^8.3.0".
warning "workspace-aggregator-b3340927-bf1a-491e-bf4a-a3ffec7e9fd4 > theia-blueprint-browser > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react@^0.14.0 || ^15.0.0 || ^16.0.0".
warning "workspace-aggregator-b3340927-bf1a-491e-bf4a-a3ffec7e9fd4 > theia-blueprint-browser > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react-dom@^0.14.0 || ^15.0.0 || ^16.0.0".
[5/5] 🔨  Building fresh packages...
[14/14] ⠠ electron-chromedriver
[-/14] ⠠ waiting...
[-/14] ⠠ waiting...
[-/14] ⠠ waiting...
error /Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver: Command failed.
Exit code: 1
Command: node ./download-chromedriver.js
Arguments: 
Directory: /Users/ragipselcuk/Sandbox/blueprint/theia-blueprint/node_modules/electron-chromedriver
Output:
node:internal/process/promises:279
            triggerUncaughtException(err, true /* fromPromise */);
            ^

GotError [HTTPError]: Response code 404 (Not Found) for https://github.com/electron/electron/releases/download/v9.0.0/**chromedriver-v9.0.0-darwin-arm64.zip**
``` 

4. How can I change the fresh package version with different way? I also tried to environmental setting change but it doesn't work.
5. I checked "[https://github.com/electron/electron/releases/](https://github.com/electron/electron/releases/)" page again and checked the version 9.0.0 and **chromedriver-v9.0.0-darwin-arm64.zip** file doesn't exist.


<img width="868" alt="Screenshot 2023-05-25 at 23 18 16" src="https://github.com/eclipse-theia/theia/assets/6311780/207f0743-989f-4a23-9e3b-4123d869feb4">



is there any suggestion ? thanks a lot

**Reply by vince-fugnitto** [reply, marked as answer]
@RagipSelcuk that isn't really how dependencies work, you can't modify stuff under `node_modules` (which is installed through `yarn`) and expect them to persist. If you need a newer version of `electron-chromedriver` then you should pull a newer version by updating the [version range](https://github.com/eclipse-theia/theia-blueprint/blob/5850adb9220e3cea711ebcf4ef89852c7720778e/applications/electron/package.json#L102) or using [yarn resolutions](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions/).

**Reply by RagipSelcuk** [reply]
Hi @vince-fugnitto,

Thank you for patiently listening to me. [version range](https://github.com/eclipse-theia/theia-blueprint/blob/5850adb9220e3cea711ebcf4ef89852c7720778e/applications/electron/package.json#L102) worked for me.

Here is what I have changed;

- Change directory to "theia-blueprint/applications/electron"
- Updated the package.json as below to "electron-chromedriver": "11.0.0",

```typescript


  "devDependencies": {
    "@theia/cli": "1.37.0",
    "@types/js-yaml": "^3.12.0",
    "@types/yargs": "17.0.7",
    "@wdio/cli": "^6.10.2",
    "@wdio/local-runner": "^6.10.2",
    "@wdio/mocha-framework": "^6.8.0",
    "@wdio/spec-reporter": "^6.8.1",
    "app-builder-lib": "22.10.5",
    "chai": "^4.2.0",
    "electron": "^23.2.4",
    "electron-builder": "23.6.0",
    "electron-chromedriver": "11.0.0",
    "electron-mocha": "^9.3.2",
    "js-yaml": "^3.12.0",
    "mocha": "^8.2.1",
    "rimraf": "^2.7.1",
    "ts-node": "^10.0.0",
    "wdio-chromedriver-service": "^6.0.4",
    "webdriverio": "^6.10.2",
    "yargs": "17.2.1",
    "concurrently": "^3.5.0"
  },

``` 

- Changed directory to "theia-blueprint"
- #yarn

I'll give information about yarn process results.

Edit: yarn process is successfully completed with electron-chromedriver": "11.0.0".

Thanks a lot.

**Comment by gwadson** [comment]
Hi @vince-fugnitto 
I experienced the same issue and had to change the Chrome driver version to 11 for the attached video to work.


https://github.com/eclipse-theia/theia/assets/68423815/a606ae5a-c0ee-43a2-aa62-519c1f51f647
